goa v2 grpc codegen panics when two methods return same collection

The code generator panics if a service has two methods that have the same result type of collection. If one or the other is changed to return Empty or a non-collection then the panic goes away.

`$GOPATH/src/foo/design/api.go`
```
package design

import (
	. "goa.design/goa/dsl"
)

var _ = Service("Foo", func() {
	Method("a", func() {
		Result(CollectionOf(Bar))

		GRPC(func() {
			Response(CodeOK)
		})
	})

	Method("b", func() {
		Result(CollectionOf(Bar))

		GRPC(func() {
			Response(CodeOK)
		})
	})
})

var Bar = ResultType("application/vnd.foo.bar", func() {
	TypeName("Bar")
	Attributes(func() {
		Field(1, "name", String)
		Required("name")
	})
	View("default", func() {
		Attribute("name")
	})
})
```
Run the following:
```
~/gopath/src/foo$ goa gen foo/design
exit status 2
panic: runtime error: index out of range

goroutine 1 [running]:
goa.design/goa/grpc/codegen.ServicesData.analyze.func1(0xc00031fe00, 0xc00032d180)
	/home/me/gopath/src/goa.design/goa/grpc/codegen/service_data.go:433 +0x1c2
goa.design/goa/grpc/codegen.ServicesData.analyze(0xc00026eb70, 0xc0002318c0, 0x3)
	/home/me/gopath/src/goa.design/goa/grpc/codegen/service_data.go:527 +0x226f
goa.design/goa/grpc/codegen.ServicesData.Get(0xc00026eb70, 0xa1837b, 0x3, 0xc000236480)
	/home/me/gopath/src/goa.design/goa/grpc/codegen/service_data.go:356 +0xb4
goa.design/goa/grpc/codegen.protoFile(0xc00024c28a, 0x7, 0xc0002318c0, 0xc0002344d8)
	/home/me/gopath/src/goa.design/goa/grpc/codegen/proto.go:29 +0x62
goa.design/goa/grpc/codegen.ProtoFiles(0xc00024c28a, 0x7, 0xef2ba0, 0x0, 0x0, 0x0)
	/home/me/gopath/src/goa.design/goa/grpc/codegen/proto.go:23 +0xda
goa.design/goa/codegen/generator.Transport(0xc00024c28a, 0x7, 0xc000229340, 0x2, 0x2, 0x4, 0x4, 0x4, 0x0, 0x0)
	/home/me/gopath/src/goa.design/goa/codegen/generator/transport.go:33 +0x5fb
goa.design/goa/codegen/generator.Generate(0x7ffc4d644a08, 0x1, 0xa1876e, 0x3, 0x0, 0x0, 0xc000232d20, 0x0, 0xffffffffffffffff)
	/home/me/gopath/src/goa.design/goa/codegen/generator/generate.go:62 +0x362
main.main()
	/home/me/gopath/src/foo/goa236066888/main.go:54 +0x275
```

--- a/grpc/codegen/proto_test.go
+++ b/grpc/codegen/proto_test.go
@@ -22,6 +22,7 @@ func TestProtoFiles(t *testing.T) {
 		{"bidirectional-streaming-rpc", testdata.BidirectionalStreamingRPCDSL, testdata.BidirectionalStreamingRPCProtoCode},
 		{"same-service-and-message-name", testdata.MessageWithServiceNameDSL, testdata.MessageWithServiceNameProtoCode},
 		{"method-with-reserved-proto-name", testdata.MethodWithReservedNameDSL, testdata.MethodWithReservedNameProtoCode},
+		{"multiple-methods-same-return-type", testdata.MultipleMethodsSameResultCollectionDSL, testdata.MultipleMethodsSameResultCollectionProtoCode},
 	}
 	for _, c := range cases {
 		t.Run(c.Name, func(t *testing.T) {

--- a/grpc/codegen/service_data.go
+++ b/grpc/codegen/service_data.go
@@ -429,8 +429,19 @@ func (d ServicesData) analyze(gs *expr.GRPCServiceExpr) *ServiceData {
 		// collect all the nested messages and return the top-level message
 		collect := func(att *expr.AttributeExpr) *service.UserTypeData {
 			msgs := collectMessages(att, sd, seen)
-			sd.Messages = append(sd.Messages, msgs...)
-			return msgs[0]
+			if len(msgs) > 0 {
+				sd.Messages = append(sd.Messages, msgs...)
+				return msgs[0]
+			}
+			// lookup message in sd.Messages
+			if ut, ok := att.Type.(expr.UserType); ok {
+				for _, t := range sd.Messages {
+					if t.Name == ut.Name() {
+						return t
+					}
+				}
+			}
+			return nil
 		}
 
 		var (

--- a/grpc/codegen/testdata/dsls.go
+++ b/grpc/codegen/testdata/dsls.go
@@ -643,3 +643,20 @@ var MethodWithReservedNameDSL = func() {
 		})
 	})
 }
+
+var MultipleMethodsSameResultCollectionDSL = func() {
+	var ResultT = ResultType("application/vnd.goa.result", func() {
+		TypeName("ResultT")
+		Field(1, "BooleanField", Boolean)
+	})
+	Service("MultipleMethodsSameResultCollection", func() {
+		Method("method_a", func() {
+			Result(CollectionOf(ResultT))
+			GRPC(func() {})
+		})
+		Method("method_b", func() {
+			Result(CollectionOf(ResultT))
+			GRPC(func() {})
+		})
+	})
+}

--- a/grpc/codegen/testdata/proto_code.go
+++ b/grpc/codegen/testdata/proto_code.go
@@ -361,3 +361,33 @@ message StringRequest {
 message StringResponse {
 }
 `
+
+const MultipleMethodsSameResultCollectionProtoCode = `
+syntax = "proto3";
+
+package multiple_methods_same_result_collection;
+
+option go_package = "multiple_methods_same_result_collectionpb";
+
+// Service is the MultipleMethodsSameResultCollection service interface.
+service MultipleMethodsSameResultCollection {
+	// MethodA implements method_a.
+	rpc MethodA (MethodARequest) returns (ResultTCollection);
+	// MethodB implements method_b.
+	rpc MethodB (MethodBRequest) returns (ResultTCollection);
+}
+
+message MethodARequest {
+}
+
+message ResultTCollection {
+	repeated ResultT field = 1;
+}
+
+message ResultT {
+	bool boolean_field = 1;
+}
+
+message MethodBRequest {
+}
+`

